### ISSUE TITLE
Type Error vs NoReduce for grounded functions waiting for grounded atoms, but called otherwise

### ISSUE THREAD

**Issue by vsbogd** [trueagi-io/hyperon-experimental#320]
## Code snippet

Consider the following grounded function example from #316:

`my-extension.py`:
```py
from hyperon.atoms import E, S, OperationAtom, ValueAtom
from hyperon.ext import *

def foo(x):
    print("foo(x={})".format(x))
    return E(S("Ctor"), ValueAtom(0.42))

@register_atoms
def my_function():
    return {'foo': OperationAtom('foo', foo, ['T', 'T'], True)}
```

`test-extension.metta`:
```
(: T Type)
(: Ctor (-> Number T))

!(extend-py! my-extension)
!(foo (Ctor 0.1))
```

## Expected behavior

`foo` is an instance of `OperationAtom` which is instantiated with `unwrap` flag set. It means `foo` expects grounded values on its input. `(Ctor 0.1)` has a type `T` but it is not grounded. Thus interpreter is expected to return type error here `(Error (foo (Ctor 0.1)) BadType)`. See https://github.com/trueagi-io/hyperon-experimental/issues/316#issuecomment-1573677533 and following discussion.

## Actual behavior

Interpreter returns `(foo (Ctor 0.1))` because `OperationAtom` finds that argument is not grounded expression and raises `NoReduceError` (see https://github.com/trueagi-io/hyperon-experimental/issues/316#issuecomment-1573486533):
```
[]
[(foo (Ctor 0.1))]
```

**Comment by Necr0x0Der** [comment]
Why returning `BadType` is expected behavior? It is a correctly typed expression. For me, the expected behavior is to keep `(foo (Ctor 0.1))` unreduced.

**Comment by Necr0x0Der** [comment]
`Grounded` is a meta-type - not type. We don't have possibility to define both types and meta-types atm.
The following code returns the error:
```
from hyperon import *

m = MeTTa()
m.register_atom('foo', OperationAtom('foo', foo, ['Grounded', 'Grounded'], True))
print(m.run('''
!(foo 1) ; OK
(: S T)
!(foo S) ; BadType
'''))
```
So, if one wants to explicitly restrict the argument to be `Grounded`, it is possible.
Interestingly, without `(: S T)`, there is no type error, seemingly, because `%Undefined%` matches against `Grounded`, although `S` is a `Symbol`. Is this a bug? I'm not sure. There is slightly a mess with intermix of types and meta-types.  It'd say it's more like a error in comparison to the `unwrap=True` case. Another example is that `(+ 1 S)` doesn't raise a error (remaining unreduced), while it is equivalent to `unwrap=True` in Python. Maybe, we need to sort out types and meta-types issue, and have a clear semantics for them. Without it, I disagree that the present issue is a bug.
In fact, it was designed as a feature. The whole idea was to have the possibility to write down any intermixture of symbolic and subsymbolic expressions in MeTTa. And we really can write right now:
```
(= (replace-S $expr $val)
   (map-atom $expr $r (if (== $r S) $val $r)))
!(replace-S (+ 1 S) 5) ; 6
```
which I'm happy about.
I know that non-reduction of `(+ 1 S)` causes some troubles for finding bugs in MeTTa programs. I believe that we discussed somewhere that `pragma` could be introduced to control such behavior (non-reduct vs error). However, for me, it's more about core library functions rather than `unwrap=True` in Python.  As for the latter, it can be easily changed by replacing `raise NoReduceError()` with `raise RuntimeError...`. But it's really the intended feature - not the bug. Apparently, if `pragma` is introduced, it should be taken into account in the Python `OperationObject` as well. If it is not introduced, one more flag can be added in addition to `unwrap` if one really wants to avoid non-reducts for their custom grounded functions (although the possibility to indicate both types and metatypes and match them separately would be preferable).
So, I vote for `pragma` and possibly for separation of types and meta-typed rather than changing the behavior of `unwrap=True` separately from other cases.

**Comment by vsbogd** [comment]
> We don't have possibility to define both types and meta-types atm.

We cannot declare it explicitly. Implicitly `OperationAtom` declares that it expects grounded Python value by setting `unwrap` flag. If it is ready to get non-grounded atoms as arguments it should drop `unwrap` flag.

**Comment by Necr0x0Der** [comment]
> We cannot declare it explicitly

We can for specific operations while declaring their concrete types.

> If it is ready to get non-grounded atoms as arguments it should drop unwrap flag.

The question is not if they are ready to get non-grounded atoms. `+` is not ready to get non-grounded atoms as well, but `(+ 1 S)` doesn't throw a type error, but remains unreduced. And this is the intended behavior. So, I still disagree that `unwrap=True` should behave differently.

**Comment by Necr0x0Der** [comment]
@vsbogd does it make to raise a separate issue on the need for `pragma`, which changes the behavior of such operations as `+` and `unwrap=True` to raise an error instead of `NoReduce`?

**Comment by vsbogd** [comment]
What would be the use-case of such pragma? If it is for debug purposes then maybe it is more convenient to have a command line argument for the interpreter.

**Comment by Necr0x0Der** [comment]
Yes. It is mostly for debugging. The command line argument might be ok, but personally it is less convenient for me since I run the interpreter from IDE using a hotkey rather than in terminal. And also this flag should be accessible from hyperonpy, so there should be a mechanism to pass it there somehow. So, it can be put to the interpreter environment in any way... OTOH, it will require `pass_metta=True` for all grounded operations, so it might be tricky to handle it on the `OperationObject` side....
In any case, there were a lot of complains about keeping expressions like `(+ 1 S)` unreduced. And I'd actually make error raising the default behavior. But I still want to be able to get back to the old behavior with `NoReduce`, and this is not for debugging, but opposite. So, if some code expects `NoReduce` instead of error, this should be stated in the code itself, and a command line flag is not that suitable in this scenario. But maybe this should be discussed within new issue. Or we can rename, repurpose and reopen this issue for this discussion

**Comment by vsbogd** [comment]
Well, if we would like to change the default behavior then it is better using pragma.

I think opening new issue will be simpler than reformatting this one but both ways are good to me.

**Comment by Necr0x0Der** [comment]
Reopening it with new name.
